### ISSUE TITLE
[redbean] Function to get raw HTTP request

### ISSUE THREAD

**Issue by vortinen** [jart/cosmopolitan#578]
Hi there! I have a use-case where it would be handy to access the full, raw HTTP request string from lua.

Would you accept a small PR that adds a function for this? If so I'll follow the instructions in `CONTRIBUTING.md`.

**Comment by pkulchenko** [comment]
@vortinen, can you describe your use case? Do you mean the exact HTTP request string that is being sent? Same string that is returned by `GetUrl`, but without re-encoding?

**Comment by vortinen** [comment]
Thanks @pkulchenko, yes I did mean the full HTTP request string. I'm interested in this for:
 1. Logging full incoming requests 'as-is', without any encoding, that can be parsed later
 2. Forwarding these on in a reverse-proxy situation, maybe with `X-Forwarded-For` etc. headers inserted.

As far as I can tell, `GetUrl` doesn't include headers (redbean 2.0.17)?
```
-> Using GetUrl
http://127.0.0.1:8080/abc?d=e
```

It looks like the original request string could be reconstructed up to header ordering with `GetUrl`, `GetBody` to avoid encoding, and `GetHeaders`. But, at least in this case, it would be neater to have something that exposes `lua_pushlstring(L, inbuf.p, hdrsize + payloadlength)` directly :)

**Comment by pkulchenko** [comment]
Yes, `GetUrl` doesn't include the headers, but there is `GetHeader` method that does.

I actually have a very similar need, as I do use redbean as the reverse proxy, so would be interested in speeding up and simplifying this processing. I do construct a full request that I then sent using `Fetch` and forward the response back. I was actually more interesting in sending the buffer returned by `Fetch` back to the client, but some of the headers still need to be modified/removed, as redbean doesn't allow them to be set.

If we're talking about even more complete solution, then we should also considering streaming responses from Fetch to avoid double buffering them, which is not yet supported either.

This is a long way to say that getting raw HTTP request may be not very useful in the current situation (as there is not much you can do with it, since you can't pass it on to `Fetch`) and the current functionality still covers the reverse-proxy scenario. Maybe we can re-purpose the ticket to capture steps needed for a more complete solution.

**Comment by jart** [comment]
We should have a function that returns a string containing the contents of the HTTP request buffer.

What we also might want to do, is provide Lua bindings for `ParseHttpMessage` so that Lua code can parse HTTP messages quickly. Then we might stand a better chance of implementing a `Fetch()` replacement in Lua code that could (a) use promises; or (b) do streaming.

**Comment by pkulchenko** [comment]
Yes, that'd be quite useful, but it will also require a way to *set* request headers for Fetch and return response headers back.

Maybe there has to be a way to facilitate specifically proxy handling by sending request to Fetch and sending the (streaming) response back.
